lib/std/interface.zig

```diff
@@ -0,0 +1,478 @@
+const std = @import("std.zig");
+const Allocator = std.mem.Allocator;
+const trait = std.meta.trait;
+const assert = std.debug.assert;
+
+pub const SelfType = @OpaqueType();
+
+fn makeSelfPtr(ptr: var) *SelfType {
+    if (comptime !trait.isSingleItemPtr(@TypeOf(ptr))) {
+        @compileError("SelfType pointer initialization expects pointer parameter.");
+    }
+
+    const T = std.meta.Child(@TypeOf(ptr));
+
+    if (@sizeOf(T) > 0) {
+        return @ptrCast(*SelfType, ptr);
+    } else {
+        return undefined;
+    }
+}
+
+fn selfPtrAs(self: *SelfType, comptime T: type) *T {
+    if (@sizeOf(T) > 0) {
+        return @alignCast(@alignOf(T), @ptrCast(*align(1) T, self));
+    } else {
+        return undefined;
+    }
+}
+
+fn constSelfPtrAs(self: *const SelfType, comptime T: type) *const T {
+    if (@sizeOf(T) > 0) {
+        return @alignCast(@alignOf(T), @ptrCast(*align(1) const T, self));
+    } else {
+        return undefined;
+    }
+}
+
+pub const Storage = struct {
+    pub const NonOwning = struct {
+        erased_ptr: *SelfType,
+
+        pub fn init(args: var) !NonOwning {
+            if (args.len != 1) {
+                @compileError("NonOwning storage expected a 1-tuple in initialization.");
+            }
+
+            return NonOwning{
+                .erased_ptr = makeSelfPtr(args.@"0"),
+            };
+        }
+
+        pub fn getSelfPtr(self: NonOwning) *SelfType {
+            return self.erased_ptr;
+        }
+
+        pub fn deinit(self: NonOwning) void {}
+    };
+
+    pub const Owning = struct {
+        allocator: *Allocator,
+        erased_ptr: *SelfType,
+
+        pub fn init(args: var) !Owning {
+            if (args.len != 2) {
+                @compileError("Owning storage expected a 2-tuple in initialization.");
+            }
+
+            const AllocT = @TypeOf(args.@"0");
+
+            var mem = try args.@"1".create(AllocT);
+            mem.* = args.@"0";
+
+            return Owning{
+                .allocator = args.@"1",
+                .erased_ptr = makeSelfPtr(mem),
+            };
+        }
+
+        pub fn getSelfPtr(self: Owning) *SelfType {
+            return self.erased_ptr;
+        }
+
+        pub fn deinit(self: Owning) void {
+            self.allocator.destroy(selfPtrAs(self.erased_ptr, u8));
+        }
```

Comment:
I don't like the `self.allocator.destroy(selfPtrAs(self.erased_ptr, u8));` line in this code. This doesn't look right to me.

Imho the better way would be to allocate a `AllocT` aligned `[]u8` with `@sizeOf(AllocT)` and store that array in `erased_ptr`, only casting the slice to `*AllocT` in `getSelfPtr`. This will pass the correct size to `shrinkFn` in the allocator.